plugins/bundle/plugin.go

```diff
@@ -0,0 +1,309 @@
+// Copyright 2018 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+// Package bundle implements bundle downloading.
+package bundle
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"net/http"
+	"time"
+
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/bundle"
+	"github.com/open-policy-agent/opa/plugins"
+	"github.com/open-policy-agent/opa/storage"
+	"github.com/open-policy-agent/opa/util"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// min amount of time to wait following a failure
+	minRetryDelay          = time.Millisecond * 100
+	defaultMinDelaySeconds = int64(60)
+	defaultMaxDelaySeconds = int64(120)
+)
+
+// PollingConfig represents configuration for the plugin's polling behaviour.
+type PollingConfig struct {
+	MinDelaySeconds *int64 `json:"min_delay_seconds,omitempty"` // min amount of time to wait between successful poll attempts
+	MaxDelaySeconds *int64 `json:"max_delay_seconds,omitempty"` // max amount of time to wait between poll attempts
+}
+
+// Config represents configuration the plguin.
+type Config struct {
+	Name    string        `json:"name"`
+	Service string        `json:"service"`
+	Polling PollingConfig `json:"polling"`
+}
+
+func (c *Config) validateAndInjectDefaults(services []string) error {
+
+	if c.Name == "" {
+		return fmt.Errorf("invalid bundle name %q", c.Name)
+	}
+
+	found := false
+
+	for _, svc := range services {
+		if svc == c.Service {
+			found = true
+			break
+		}
+	}
+
+	if !found {
+		return fmt.Errorf("invalid service name %q in bundle %q", c.Service, c.Name)
+	}
+
+	// Reject bad min/max values
+	if c.Polling.MaxDelaySeconds != nil && c.Polling.MinDelaySeconds != nil {
+		if *c.Polling.MaxDelaySeconds < *c.Polling.MinDelaySeconds {
+			return fmt.Errorf("max polling delay must be >= min polling delay in bundle %q", c.Name)
+		}
+	}
+
+	var min, max int64
+
+	if c.Polling.MaxDelaySeconds == nil {
+		max = defaultMaxDelaySeconds
+	} else {
+		max = *c.Polling.MaxDelaySeconds
+	}
+
+	if c.Polling.MinDelaySeconds == nil {
+		min = defaultMinDelaySeconds
+	} else {
+		min = *c.Polling.MinDelaySeconds
+	}
+
+	if max < min {
+		if c.Polling.MinDelaySeconds != nil {
+			max = min
+		} else {
+			min = max
+		}
+	}
```

Comment:
For ex. if user sets MinDelaySeconds = 150 and does not set MaxDelaySeconds. Given,
defaultMinDelaySeconds = int64(60) and defaultMaxDelaySeconds = int64(120).

In this case, min = 150 and max = 150. Hence  c.Polling.MaxDelaySeconds and c.Polling.MinDelaySeconds  both are 150.

Hence **delay = time.Duration(((max - min) * rand.Float64()) + min)** = 150.


IMO when you update *max=min* in line 86, you should update *min* as well. Same applies to line 88.

Reply:
I'm going to simplify this and require that the user supply both min and max if they do at all. Then there's no implicit behaviour. Since these are supplied via config files (or API objects in the future) I'm not worried about ease of use.